enforcer-rules/src/main/java/org/apache/maven/plugins/enforcer/Descriptors.java

```diff
@@ -0,0 +1,148 @@
+import org.codehaus.plexus.component.configurator.ComponentConfigurator;
+import org.codehaus.plexus.configuration.PlexusConfiguration;
+import org.codehaus.plexus.configuration.xml.XmlPlexusConfiguration;
+import org.codehaus.plexus.util.xml.Xpp3DomBuilder;
+import org.codehaus.plexus.util.xml.pull.XmlPullParserException;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
+
+/**
+ * An enforcer rule that will invoke rules from an external resource
+ *
+ * @author <a href="mailto:gastaldi@apache.org">George Gastaldi</a>
+ */
+public class Descriptors extends AbstractNonCacheableEnforcerRule
+{
+    String descriptorRef;
+
+    String descriptor;
+
+    @Override
+    public void execute( EnforcerRuleHelper helper ) throws EnforcerRuleException
+    {
+        // Find descriptor
+        EnforcerDescriptor enforcerDescriptor = getEnforcerDescriptor( helper );
+        for ( EnforcerRule rule : enforcerDescriptor.getRules() )
+        {
+            rule.execute( helper );
+        }
+    }
+
+    /**
+     * Resolve the {@link EnforcerDescriptor} based on the provided {@link #descriptor} or {@link #descriptorRef}
+     *
+     * @param helper used to build the {@link EnforcerDescriptor}
+     * @return an {@link EnforcerDescriptor} for this rule
+     * @throws EnforcerRuleException if any failure happens while reading the descriptor
+     */
+    EnforcerDescriptor getEnforcerDescriptor( EnforcerRuleHelper helper )
+            throws EnforcerRuleException
+    {
+        try ( InputStream descriptorStream = resolveDescriptor( helper ) )
+        {
+            EnforcerDescriptor descriptor = new EnforcerDescriptor();
+            // To get configuration from the enforcer-plugin mojo do:
+            //helper.evaluate(helper.getComponent(MojoExecution.class).getConfiguration().getChild("fail").getValue())
+            // Get the enforcer plugin's class resolver
+            ClassRealm realm = helper.getComponent( MojoExecution.class ).getMojoDescriptor().getRealm();
+            ComponentConfigurator configurator = helper.getComponent( ComponentConfigurator.class, "basic" );
+            // Configure EnforcerDescriptor from the XML
+            configurator.configureComponent( descriptor, toPlexusConfiguration( descriptorStream ), helper, realm );
+            return descriptor;
+        }
+        catch ( EnforcerRuleException e )
+        {
+            throw e;
+        }
+        catch ( Exception e )
+        {
+            throw new EnforcerRuleException( "Error while enforcing rules", e );
+        }
+    }
+
+    private InputStream resolveDescriptor( EnforcerRuleHelper helper ) throws EnforcerRuleException
+    {
+        InputStream descriptorStream;
+        if ( descriptorRef != null )
+        {
+            descriptorStream = Thread.currentThread().getContextClassLoader()
+                    .getResourceAsStream( "enforcer-rules/" + descriptorRef + ".xml" );
```

Comment:
I find this a bit lacking on self-explainability. I can imagine I'll need to google up every time what is the name of the magic folder. Can't we simply use a single xml element for both files and classpath resources, something like:

```
<descriptors>
    <descriptor>file:foo/bar.xml</descriptor><!-- relative to ${basedir} -->
    <descriptor>file:${maven.multiModuleProjectDirectory}/bar/baz.xml</descriptor><!-- absolute path within the source tree -->
    <descriptor>classpath:foo/bar.xml</descriptor><!-- class path resource -->
</descriptors>
```

Reply:
That's an interesting approach. Just out of curiosity: Do we have other maven plugins doing the same? Just thinking how other plugins managed to solve the same problem

Reply:
I prefer doing it that way in my tools, but I do not know from top of my head whether any ASF plugins do it like that. 

If you want to stay similar with Assembly plugin, I can live with it.

Maybe we could try to find a better name for the top level element. How about `<externalRules>`?

Reply:
+1, that sounds better, I suck at naming things 😀

Reply:
> If you want to stay similar with Assembly plugin, I can live with it.

I don't mind changing it, I just want to avoid potential CVEs in the future 😀

Reply:
@ppalaga now it's called `location` and it expects a file path or a resource from the classpath if it starts with `classpath:`
```xml
<externalRules>
    <location>classpath:enforcer-rules/quarkus.xml</location>
</externalRules>
```

You don't specify the `classpath:` prefix, it will respect the implementation in `org.apache.maven.plugin.PluginParameterExpressionEvaluator#alignToBaseDirectory`:

```xml
<externalRules>
    <location>classpath:enforcer-rules/quarkus.xml</location>
</externalRules>
```